x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -291,20 +291,46 @@ public Automaton allowedActionsMatcher(String index) {
+                            if (fieldPermissions == null) {
+                                // Most indices rely on the default (empty) field permissions object, so we optimize for that case
+                                // Using an immutable single item set is significantly faster because it avoids any of the hashing
+                                // and backing set creation.
+                                fieldPermissions = Set.of(group.getFieldPermissions());
+                                fieldPermissionsByIndex.put(index, fieldPermissions);
+                            } else if (fieldPermissions.size() == 1) {
+                                FieldPermissions fp = group.getFieldPermissions();
+                                if (fieldPermissions.contains(fp) == false) {
+                                    // This index doesn't have a single field permissions object, replace the singleton with a real Set
+                                    final FieldPermissions existing = fieldPermissions.iterator().next();
+                                    fieldPermissions = new HashSet<>();
+                                    fieldPermissions.add(existing);
+                                    fieldPermissions.add(fp);
+                                    fieldPermissionsByIndex.put(index, fieldPermissions);
+                                }
+                            } else {
+                                fieldPermissions.add(group.getFieldPermissions());
+                            }
+
+                            DocumentLevelPermissions docPermissions;
                             if (group.hasQuery()) {
-                                permissions.addAll(group.getQuery());
+                                docPermissions = roleQueriesByIndex.computeIfAbsent(index, (k) -> new DocumentLevelPermissions());
+                                docPermissions.addAll(group.getQuery());
                             } else {
                                 // if more than one permission matches for a concrete index here and if
                                 // a single permission doesn't have a role query then DLS will not be
                                 // applied even when other permissions do have a role query
-                                permissions.setAllowAll();
+                                docPermissions = DocumentLevelPermissions.ALLOW_ALL;
+                                // don't worry about what's already there - just overwrite it, it avoids doing a 2nd hash lookup.
+                                roleQueriesByIndex.put(index, docPermissions);
+                            }
+
+                            if (index.equals(indexOrAlias) == false) {
+                                fieldPermissionsByIndex.put(indexOrAlias, fieldPermissions);
+                                roleQueriesByIndex.put(indexOrAlias, docPermissions);
```

Comment:
Really a note to my(future)self: The original code uses a `roleQueriesByIndex.putIfAbsent` but I think the `put` here is functionally equivalent because:
* If `indexOrAlias` is an alias, it must haven't got a value yet because `requestedIndicesOrAliases` is a `Set`.
* If `indexOrAlias` is a concrete index, the `if` condition evalutates to `false` and this code won't run.

Reply:
That this has changed is an accident - I simply didn't notice. 
`fieldPermissionsByIndex` was a `put` but `roleQueriesByIndex` was `putIfAbsent`, however there's no real reason for those to be different.